[Site Isolation] Text in cross-site iframe isn't translated if frame is added after user translates page

RupinMittal opened the pull request:
#### a295ea495b3bdefa3963df88c83e7b0bdf5d1449
<pre>
[Site Isolation] Text in cross-site iframe isn&apos;t translated if frame is added after user translates page
<a href="https://bugs.webkit.org/show_bug.cgi?id=301981">https://bugs.webkit.org/show_bug.cgi?id=301981</a>
<a href="https://rdar.apple.com/164059026">rdar://164059026</a>

Reviewed by Wenson Hsieh and Sihui Liu.

If the user translates a page, and later a cross-site iframe is added, the
text in this iframe should be translated automatically as well. This happens
with site isolation off, but not with site isolation on.

With site isolation off, the process is:
1. User translates the page
2. WebPageProxy::startTextManipulations sends IPC to WebPage::startTextManipulations
3. WebPage stores m_textManipulationIncludesSubframes and textManipulationExclusionRules
4. Cross-site iframe is added
5. WebPage::didCommitLoad calls startTextManipulationForFrame because
   m_textManipulationIncludesSubframes is already set
6. Text gets translated

With site isolation on, the iframe is added to a new WebPage instead of the
existing one. So in step 5, WebPage::didCommitLoad doesn&apos;t call
startTextManipulationForFrame because this new WebPage doesn&apos;t have
m_textManipulationIncludesSubframes set. And so the text isn&apos;t translated.

To fix this, WebPageProxy::startTextManipulations stores these
TextManipulationParameters so that they&apos;ll be sent to new WebPages created
in the future (via WebPageCreationParamters). Once WebPage::didCommitLoad
is called, it&apos;ll call startTextManipulationForFrame.

These TextManipulationParameters are cleared from WebPageProxy whenever the
main frame navigates.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/TextManipulationParameters.h: Added.
* Source/WebKit/Shared/TextManipulationParameters.serialization.in: Added.
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):

If the main frame navigates, clear the TextManipulationParameters.

(WebKit::WebPageProxy::creationParameters):

Send the TextManipulationParameters to the new WebPage.

(WebKit::WebPageProxy::startTextManipulations):

Store the TextManipulationParameters.

* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, TextManipulationInIframeIfIframeIsAddedAfterTranslationCall)):

Canonical link: <a href="https://commits.webkit.org/302630@main">https://commits.webkit.org/302630@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/222b29f95f8c4dcefa8d2b3647474274d325d9ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129706 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1967 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40562 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137094 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81169 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fbccce7f-b03d-4784-945c-ffb678aaeffb) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131577 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1915 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1856 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98807 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66635 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/becf5d65-89b5-4557-a09f-1bc475e45688) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132653 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1466 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116171 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79484 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/49c763c9-0f18-4cf1-a8a5-a80c9ac05d11) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1384 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34304 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80367 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109861 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34806 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139576 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1761 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1658 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107318 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1806 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107190 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1422 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31017 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54511 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20239 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1834 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65203 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1648 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1683 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1755 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/Shared/TextManipulationParameters.h:
```diff
@@ -0,0 +1,38 @@
+namespace WebKit {
+
+struct TextManipulationParameters {
+    bool includeSubframes;
```
Small nit - let's give this a default value, to guard against UB in the future.

```suggestion
    bool includeSubframes { false };
```

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -5150,6 +5150,49 @@ HTTPServer server({
+    }];
+    Util::run(&done);
+
+    [webView evaluateJavaScript: @"addCrossDomainIframe();" completionHandler:nil];
```
Another small style nit -

```suggestion
    [webView evaluateJavaScript:@"addCrossDomainIframe();" completionHandler:nil];
```

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7536,6 +7536,11 @@ void WebPageProxy::didCommitLoadForFrame(IPC::Connection& connection, FrameIdent
 #endif
+
+    if (frame->isMainFrame()) {
+        m_textManipulationItemCallback = nullptr;
```
Why do we need to make this change? The current behavior is to keep translation on after same-site navigation (e.g. if you click a link in apple.com, the page will still be translated).

RupinMittal replied:
Ok, just removed this.

RupinMittal committed:
[Site Isolation] Text in cross-site iframe isn't translated if frame is added after user translates page
https://bugs.webkit.org/show_bug.cgi?id=301981
rdar://164059026

Reviewed by Wenson Hsieh and Sihui Liu.

If the user translates a page, and later a cross-site iframe is added, the
text in this iframe should be translated automatically as well. This happens
with site isolation off, but not with site isolation on.

With site isolation off, the process is:
1. User translates the page
2. WebPageProxy::startTextManipulations sends IPC to WebPage::startTextManipulations
3. WebPage stores m_textManipulationIncludesSubframes and textManipulationExclusionRules
4. Cross-site iframe is added
5. WebPage::didCommitLoad calls startTextManipulationForFrame because
   m_textManipulationIncludesSubframes is already set
6. Text gets translated

With site isolation on, the iframe is added to a new WebPage instead of the
existing one. So in step 5, WebPage::didCommitLoad doesn't call
startTextManipulationForFrame because this new WebPage doesn't have
m_textManipulationIncludesSubframes set. And so the text isn't translated.

To fix this, WebPageProxy::startTextManipulations stores these
TextManipulationParameters so that they'll be sent to new WebPages created
in the future (via WebPageCreationParamters). Once WebPage::didCommitLoad
is called, it'll call startTextManipulationForFrame.

These TextManipulationParameters are cleared from WebPageProxy whenever the
main frame navigates.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/TextManipulationParameters.h: Added.
* Source/WebKit/Shared/TextManipulationParameters.serialization.in: Added.
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):

If the main frame navigates, clear the TextManipulationParameters.

(WebKit::WebPageProxy::creationParameters):

Send the TextManipulationParameters to the new WebPage.

(WebKit::WebPageProxy::startTextManipulations):

Store the TextManipulationParameters.

* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, TextManipulationInIframeIfIframeIsAddedAfterTranslationCall)):

Canonical link: https://commits.webkit.org/302630@main
